### PR TITLE
feat: Add WinDSR support

### LINKED ISSUE #4079: Enable DSR support in Windows nodes.
**Describe the request**
kubeproxy since 1.19  with Windows 10C supports DSR now. To support Calico Windows, we need to allow users to enable DSR.

**Explain why AKS Engine needs it**
Support Calico Windows.

**Describe the solution you'd like**
1. Add `EnableWindowsDSR` in WindowsProfile with default value `false`. We can consider to change the default value to `true` later.

2. In the file `parts/k8s/kuberneteswindowssetup.ps1`

```
# DSR
$global:EnableDSR = [System.Convert]::ToBoolean("{{WrapAsVariable "enableWindowsDSR" }}");
```

3. In method `Write-KubeClusterConfig` in the file `parts/k8s/kuberneteswindowsfunctions.ps1`
```

    $Global:ClusterConfiguration | Add-Member -MemberType NoteProperty -Name Kubernetes -Value @{
        ...
        Network      = @{
            ServiceCidr = $global:KubeServiceCIDR;
            ClusterCidr = $global:KubeClusterCIDR;
            DnsIp       = $KubeDnsServiceIp;
            EnableDSR   = $global:EnableDSR
        };
        ...
    }
```

4. In the file `staging/provisioning/windows/kubeproxystart.ps1`
```
$features = @()

# add dualstack feature gate if dualstack enabled
$isDualStackEnabled = ("IPv6DualStack=true" | ? { $Global:ClusterConfiguration.Kubernetes.Kubelet.ConfigArgs -match $_ }) -ne $null
if ($isDualStackEnabled) {
   $features += @("'IPv6DualStack=true'")
}

# add dsr feature gate if dsr enabled
if ($Global:ClusterConfiguration.Kubernetes.Network.EnableDSR) {
    $features += @("'WinOverlay=false'")
    $features += @("'WinDSR=true'")
    $global:KubeproxyArgList += @("--enable-dsr=true")
}

if ($features.Count -gt 0) {
    $global:KubeproxyArgList += @("--feature-gates=" + ($features -join ","))
}
```

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by AbelHu**
**Reason for Change**:

**Issue Fixed**:

Fixes #4079

**Credit Where Due**:

Does this change contain code from or inspired by another project?

- [x] No
- [ ] Yes

If "Yes," did you notify that project's maintainers and provide attribution?

- [ ] No
- [ ] Yes

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [] includes documentation
- [x] adds unit tests
- [] tested upgrade from previous version

### CODE REVIEW

**Comment by AbelHu** [comment]
@marosset @jsturtevant @mainred Please help to take a look at this PR. Expect your comments for the code logic. I will add unit tests and do more testing after you confirm the code logic.

File: pkg/api/types.go [hunk 1]
```diff
@@ -1140,6 +1141,22 @@ func (p *Properties) GetKubeProxyFeatureGates() string {
...
+	featureGates := ""
+	if p.FeatureFlags.IsFeatureEnabled("EnableIPv6DualStack") {
+		featureGates += "'IPv6DualStack=true'"
+	}
+	if p.FeatureFlags.IsFeatureEnabled("EnableWinDSR") {
+		if featureGates != "" {
+			featureGates += ","
+		}
+		// WinOverlay must be set to false
+		featureGates += "'WinDSR=true','WinOverlay=false'"
```
Comment (marosset) [inline pkg/api/types.go:1155]: Is WinOverlay something users can set in the apimodel?
If so we should add a check in validation.go and warn/fail if both WinOverlay and WinDSR are set and they are mutally exclusive.

**Review by marosset — COMMENTED** [review]
Changes / logic look good so far.
Adding the kube-proxy flags to KubeClusterCongih is the right thing to do here.

File: pkg/api/types.go (hunk 1 above)
Comment (jsturtevant) [reply pkg/api/types.go:1155]: Doesn't look like we had the ability to add features gates to the Windows kube-proxy prior to this.

Comment (AbelHu) [reply pkg/api/types.go:1155]: Agree with @jsturtevant, there is no other way to set feature gates to the Windows kube-proxy.

**Comment by AbelHu** [comment]
@marosset @jsturtevant @mainred I have updated the PR and it is ready for code review. Please help to take a look. Thanks.

**Review by mainred — COMMENTED** [review]
Do we need a doc for this,  it looks good to me, but let upstream folks decide.

**Comment by AbelHu** [comment]
> s, it looks good to me, but let upstream folks decide.

Thanks for reminder. I added a doc for `enableWinDSR`

File: pkg/api/types_test.go [hunk 2]
```diff
@@ -5284,6 +5284,14 @@ func TestIsFeatureEnabled(t *testing.T) {
 			},
 			expected: false,
 		},
+		{
+			name:    "wind dsr",
```
Comment (jsturtevant) [inline pkg/api/types_test.go:5288]: nit: should this be `windows dsr`?

File: pkg/api/types_test.go [hunk 3]
```diff
@@ -5763,6 +5771,61 @@ func TestGetKubeProxyFeatureGates(t *testing.T) {
...
+		{
+			name: "both IPV6 and WinDSR enabled",
+			properties: &Properties{
+				FeatureFlags: &FeatureFlags{
+					EnableIPv6DualStack: true,
+					EnableWinDSR:        true,
+				},
+			},
+			expectedFeatureGates: "\"IPv6DualStack=true\", \"WinDSR=true\", \"WinOverlay=false\"",
+		},
```
Comment (jsturtevant) [inline pkg/api/types_test.go:5814]: Is it worth adding a test here for other flags that are not windows related don't set the value?

File: pkg/engine/template_generator_test.go [hunk 4]
```diff
@@ -1457,6 +1458,17 @@ func TestTemplateGenerator_FunctionMap(t *testing.T) {
...
+		{
+			Name:     "GetKubeProxyFeatureGatesPsh",
+			FuncName: "GetKubeProxyFeatureGatesPsh",
+			MutateFunc: func(cs api.ContainerService) api.ContainerService {
+				cs.Properties.FeatureFlags = &api.FeatureFlags{
+					EnableWinDSR: true,
+				}
+				return cs
+			},
+			ExpectedResult: "\"WinDSR=true\", \"WinOverlay=false\"",
```
Comment (jsturtevant) [inline pkg/engine/template_generator_test.go:1470]: Is there a test here for IpV6 on windows?

File: staging/provisioning/windows/kubeproxystart.ps1 [hunk 5]
```diff
@@ -17,10 +18,13 @@ while (!$hnsNetwork) {
...
-$isDualStackEnabled = ("IPv6DualStack=true" | ? { $Global:ClusterConfiguration.Kubernetes.Kubelet.ConfigArgs -match $_ }) -ne $null
-if ($isDualStackEnabled) {
-    $global:KubeproxyArgList += @("--feature-gates=IPv6DualStack=true")
+# enable WinDsr if WinDsr feature gate is enabled
+if ($KubeproxyFeatureGates -contains "WinDSR=true") {
+    $global:KubeproxyArgList += @("--enable-dsr=true")
+}
+
+if ($KubeproxyFeatureGates.Count -ne 0) {
+    $global:KubeproxyArgList += @("--feature-gates='" + ($KubeproxyFeatureGates -join "','") + "'")
```
Comment (jsturtevant) [inline staging/provisioning/windows/kubeproxystart.ps1:27]: Isn't `kubeproxyFeatureGates` already in in the format of "key=value," from `GetKubeProxyFeatureGatesWindowsArguments` (https://github.com/Azure/aks-engine/pull/4104/files#diff-dfee74d02fd350fb1ee1ffbbbc32acf4cb075c9ae42b6e31c8c7acefe7fed623R1164)

**Comment by jsturtevant** [comment]
@AbelHu do you know if we can use DSR with azure cni or another cni other than Calico?  It would be nice to run tests against this configuration while Calico support is in the works.

File: staging/provisioning/windows/kubeproxystart.ps1 (hunk 5 above)
Comment (AbelHu) [reply staging/provisioning/windows/kubeproxystart.ps1:27]: $KubeproxyFeatureGates is a list in https://github.com/Azure/aks-engine/pull/4104/files?file-filters%5B%5D=.go&file-filters%5B%5D=.ps1#diff-bb0963ea67ff5314b6c48f54d3bb555f57341f0adc5f4d33f05e536214eda1e6R106. I followed the  same logic as `KubeletConfigArgs`

**Comment by AbelHu** [comment]
> @AbelHu do you know if we can use DSR with azure cni or another cni other than Calico? It would be nice to run tests against this configuration while Calico support is in the works.

From my understanding, DSR is required by calico but DSR does not require Calico. I am asking the same question whether we can enable DSR by default in Windows nodes when k8s >= 1.20

**Comment by marosset** [comment]
/lgtm
@ksubrmnn @daschott can you help answer the above questions for DSR support?

**Comment by marosset** [comment]
Spoke with @jsturtevant and we can add testing in a later PR since we want to unblock calico support.

**Review by marosset — APPROVED** [review]

**Comment by daschott** [comment]
I think we should have more end-user validation before making it default.

**Pull request merged**